velox/jni/cpp/jni/JniCommon.h

```diff
@@ -0,0 +1,180 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <jni.h>
+
+#include <atomic>
+#include <iostream>
+#include <memory>
+#include "common/base/Exceptions.h"
+
+/*
+ * This header provides a collection of utility functions and classes to
+ * facilitate Java Native Interface (JNI) operations in C++. It includes
+ * mechanisms to manage global references to Java classes, convert Java types to
+ * C++ types, and manage shared pointers in conjunction with JNI calls.
+ *
+ * The `SharedPtrHandle` struct is a utility to wrap `std::shared_ptr` handling
+ * within JNI code, with optional reference counting and debug output to track
+ * object lifetimes and instance counts.
+ *
+ * Additionally, the file provides functions to:
+ * - Create global references to Java classes from their names.
+ * - Retrieve method IDs for instance and static methods from a class.
+ * - Convert Java byte arrays and strings to their C++ equivalents.
+ * - Convert Java string arrays to a vector of C++ strings.
+ *
+ * Debugging support is conditionally included to trace the creation and
+ * destruction of `SharedPtrHandle` instances, which can aid in identifying
+ * memory management issues.
+ */
+
+static jint jniVersion = JNI_VERSION_1_8;
+
+#define JNI_METHOD_SIG(name, signature) \
+  (JNINativeMethod{                     \
+      const_cast<char*>(#name),         \
+      const_cast<char*>(#signature),    \
+      reinterpret_cast<void*>(name)})
+
+struct SharedPtrHandle {
+  static std::atomic<int> global_instance_count;
+
+  std::shared_ptr<void> plan_node;
+
+  SharedPtrHandle() {
+#ifdef VELOX_ENABLE_JNI_BINDINGS_REF_DEBUG
+    global_instance_count.fetch_add(1, std::memory_order_relaxed);
+    print_instance_count("Default constructor");
+#endif VELOX_ENABLE_JNI_BINDINGS_REF_DEBUG
+  }
+
+  template <typename T>
+  SharedPtrHandle(std::shared_ptr<T> node)
+      : plan_node(std::static_pointer_cast<void>(node)) {
+#ifdef VELOX_ENABLE_JNI_BINDINGS_REF_DEBUG
+    global_instance_count.fetch_add(1, std::memory_order_relaxed);
+    print_instance_count("Constructor with node");
+#endif VELOX_ENABLE_JNI_BINDINGS_REF_DEBUG
+  }
+
+  SharedPtrHandle(const SharedPtrHandle&) {
+#ifdef VELOX_ENABLE_JNI_BINDINGS_REF_DEBUG
+    global_instance_count.fetch_add(1, std::memory_order_relaxed);
+    print_instance_count("Copy constructor");
+#endif VELOX_ENABLE_JNI_BINDINGS_REF_DEBUG
+  }
+
+  SharedPtrHandle(SharedPtrHandle&&) noexcept {
+#ifdef VELOX_ENABLE_JNI_BINDINGS_REF_DEBUG
+    global_instance_count.fetch_add(1, std::memory_order_relaxed);
+    print_instance_count("Move constructor");
+#endif VELOX_ENABLE_JNI_BINDINGS_REF_DEBUG
+  }
```

Comment:
Where do handles get moved / copied? Do the implementations really needed?

Reply:
it's usered to analyze memory leak